hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/StorageBasedLockProvider.java

```diff
@@ -0,0 +1,581 @@
+  private transient Thread shutdownThread = null;
+
+  @GuardedBy("this")
+  private StorageLockFile currentLockObj = null;
+  @GuardedBy("this")
+  private boolean isClosed = false;
+
+  private synchronized void setLock(StorageLockFile lockObj) {
+    if (lockObj != null && !Objects.equals(lockObj.getOwner(), this.ownerId)) {
+      throw new HoodieLockException("Owners do not match. Current lock owner: " + this.ownerId + " lock path: "
+          + this.lockFilePath + " owner: " + lockObj.getOwner());
+    }
+    this.currentLockObj = lockObj;
+  }
+
+  /**
+   * Default constructor for StorageBasedLockProvider, required by LockManager
+   * to instantiate it using reflection.
+   * 
+   * @param lockConfiguration The lock configuration, should be transformable into
+   *                          StorageBasedLockConfig
+   * @param conf              Storage config, ignored.
+   */
+  public StorageBasedLockProvider(final LockConfiguration lockConfiguration, final StorageConfiguration<?> conf) {
+    StorageBasedLockConfig config = new StorageBasedLockConfig.Builder()
+        .fromProperties(lockConfiguration.getConfig()).build();
+    heartbeatInterval = config.getHeartbeatPoll();
+    lockValidity = config.getLockValidityTimeout();
+
+    String configuredLocksLocation = config.getLocksLocation();
+
+    // If not configured, recalculate the locks location as .hoodie/.locks;
+    // otherwise (the lock location is configured), the configuration location is used as the folder
+    // to which the lock file is written to, and the lock file name is determined by the table's base path
+    String locksLocation = StringUtils.isNullOrEmpty(configuredLocksLocation)
+        ? String.format("%s%s%s", config.getHudiTableBasePath(), StoragePath.SEPARATOR, LOCKS_FOLDER_NAME)
+        : configuredLocksLocation;
+
+    URI uri = parseURI(locksLocation);
+    bucketName = uri.getHost(); // For most schemes, the bucket/container is the host.
+    String folderName = uri.getPath(); // Path after the bucket/container.
+
+    String fileName = StringUtils.isNullOrEmpty(configuredLocksLocation)
+        ? DEFAULT_TABLE_LOCK_FILE_NAME
+        : slugifyLockFolderFromBasePath(config.getHudiTableBasePath());
+
+    lockFilePath = buildLockObjectPath(folderName, fileName);
+    ownerId = UUID.randomUUID().toString();
+    this.logger = LOGGER;
+    this.heartbeatManager = new LockProviderHeartbeatManager(
+        ownerId,
+        heartbeatInterval,
+        this::renewLock);
+
+    try {
+      this.lockService = (StorageLock) ReflectionUtils.loadClass(
+          getLockServiceClassName(uri.getScheme()),
+          new Class<?>[] { String.class, String.class, String.class, Properties.class },
+          new Object[] { ownerId, bucketName, lockFilePath, lockConfiguration.getConfig() });
+    } catch (Throwable e) {
+      throw new HoodieLockException("Failed to load and initialize StorageLock", e);
+    }
+    shutdownThread = new Thread(() -> shutdown(true));
```

Comment:
in the shutdown thread, should we just try to do the utmost minimally required thing..
for eg, call tryExpireCurrentLock only.

Reply:
and lets be conscious to not throw any exception for shutdown hook. 
for eg, I see we are validating some states within 
```
shutdown(boolean fromShutdownHook) {
```

So, lets be conscious of that

Reply:
change this to best effort, and switch out the unlock retry call to not have retries

Reply:
done